[FLINK-25577][docs] Update GCS documentation (#18430)
apache/flink: docs/content/docs/deployment/filesystems/gcs.md

```

Note that these examples are *not* exhaustive and you can use GCS in other places as well, including your [high availability setup]({{< ref "docs/deployment/ha/overview" >}}) or the [EmbeddedRocksDBStateBackend]({{< ref "docs/ops/state/state_backends" >}}#the-rocksdbstatebackend); everywhere that Flink expects a FileSystem URI.

### GCS File System plugin

Flink provides the `flink-gs-fs-hadoop` file system to write to GCS.
This implementation is self-contained with no dependency footprint, so there is no need to add Hadoop to the classpath to use it.

`flink-gs-fs-hadoop` registers a `FileSystem` wrapper for URIs with the *gs://* scheme. It uses Google's [gcs-connector](https://mvnrepository.com/artifact/com.google.cloud.bigdataoss/gcs-connector) Hadoop library to access GCS. It also uses Google's [google-cloud-storage](https://mvnrepository.com/artifact/com.google.cloud/google-cloud-storage) library to provide `RecoverableWriter` support. 

> This file system supports the [StreamingFileSink]({{< ref "docs/connectors/datastream/streamfile_sink" >}}) and the [FileSink]({{< ref "docs/connectors/datastream/file_sink" >}}).

Reviewer: These two references are invalid, as CI complained. They are removed / renamed in FLINK-20188. I think we should now refers to "docs/content/docs/connectors/datastream/filesystem.md"

Author: Updated in https://github.com/apache/flink/pull/18430/commits/4d4b6ef9bf6c75d3dbbe3d64b58eb5ffcc218ebf.

To use `flink-gs-fs-hadoop`, copy the JAR file from the `opt` directory to the `plugins` directory of your Flink distribution before starting Flink, i.e.

```bash
mkdir ./plugins/gs-fs-hadoop
cp ./opt/flink-gs-fs-hadoop-{{< version >}}.jar ./plugins/gs-fs-hadoop/
```

### Configuration

> The underlying Hadoop file system can be [configured using Hadoop's gs configuration keys](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/CONFIGURATION.md) by adding the configurations to your `flink-conf.yaml`.

Reviewer: Correct me if I'm wrong, I think the precise description should be:
```suggestion
The underlying Hadoop file system can be [configured using gcs-connector's Hadoop configuration keys](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/CONFIGURATION.md) by adding the configurations to your `flink-conf.yaml`.
```

Author: Yes, that's better. Good catch. Updated in https://github.com/apache/flink/pull/18430/commits/4d4b6ef9bf6c75d3dbbe3d64b58eb5ffcc218ebf.

> For example, Hadoop has a `fs.gs.http.connect-timeout` configuration key. If you want to change it, you need to set `gs.http.connect-timeout: xyz` in `flink-conf.yaml`. Flink will internally translate this back to `fs.gs.http.connect-timeout`. There is no need to pass configuration parameters using Hadoop's XML configuration files.

Reviewer: ```suggestion
For example, gcs-connector has a `fs.gs.http.connect-timeout` configuration key. If you want to change it, you need to set `gs.http.connect-timeout: xyz` in `flink-conf.yaml`. Flink will internally translate this back to `fs.gs.http.connect-timeout`. There is no need to pass configuration parameters using Hadoop's XML configuration files.
```

Author: Updated in https://github.com/apache/flink/pull/18430/commits/4d4b6ef9bf6c75d3dbbe3d64b58eb5ffcc218ebf.

|-------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | gs.writer.temporary.bucket.name           | If this property is not set, temporary blobs for in-progress writes via `RecoverableWriter` will be written to same bucket as the final file being written, prefixed with `.inprogress/`. <br><br>Set this property to choose a different bucket to hold the temporary blobs. It is recommended to choose a separate bucket in order to [assign it a TTL](https://cloud.google.com/storage/docs/lifecycle), to provide a mechanism to clean up orphaned blobs that can occur when restoring from check/savepoints.<br><br>If you do use a separate bucket with a TTL for temporary blobs, attempts to restart jobs from check/savepoints after the TTL interval expires may fail. 

Reviewer: Maybe swap the first and second paragraph?

Author: Done in https://github.com/apache/flink/pull/18430/commits/4d4b6ef9bf6c75d3dbbe3d64b58eb5ffcc218ebf.

```

Note that these examples are *not* exhaustive and you can use GCS in other places as well, including your [high availability setup]({{< ref "docs/deployment/ha/overview" >}}) or the [EmbeddedRocksDBStateBackend]({{< ref "docs/ops/state/state_backends" >}}#the-rocksdbstatebackend); everywhere that Flink expects a FileSystem URI.

### GCS File System plugin

Flink provides the `flink-gs-fs-hadoop` file system to write to GCS.
This implementation is self-contained with no dependency footprint, so there is no need to add Hadoop to the classpath to use it.

`flink-gs-fs-hadoop` registers a `FileSystem` wrapper for URIs with the *gs://* scheme. It uses Google's [gcs-connector](https://mvnrepository.com/artifact/com.google.cloud.bigdataoss/gcs-connector) Hadoop library to access GCS. It also uses Google's [google-cloud-storage](https://mvnrepository.com/artifact/com.google.cloud/google-cloud-storage) library to provide `RecoverableWriter` support. 

> This file system supports being used in [file sinks]({{< ref "docs/content/docs/connectors/datastream/filesystem.md" >}}).

Reviewer 2: We've recently renamed the File Sink to FileSystem. 

```suggestion
This file system can be used with the [FileSystem connector]({{< ref "docs/content/docs/connectors/datastream/filesystem.md" >}}).
```

Author: Fixed in https://github.com/apache/flink/pull/18430/commits/01d94ab245f6baff38f2bfcfa3ef43d52a5b2bdf.

> This file system supports being used in [file sinks]({{< ref "docs/content/docs/connectors/datastream/filesystem.md" >}}).

Author: Fixed in https://github.com/apache/flink/pull/18430/commits/de782534bf4fb67f9d308e0d47281d2a24e18d13. Thanks for the hint on how to build locally.
